### PR TITLE
Cleanup/optimize many string.Compare overloads, part 1

### LINKED ISSUE #6465: string.Compare unintentionally doing O(n) compare
`string.Compare` has some logic to short-circuit if the strings it's comparing are equal, and the indices are equal. Except, it looks like it's calling String's overload of `==` instead of a reference comparison: https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/String.cs#L2124

This probably hasn't been discovered before since if 2 strings are ordinally equal, then they should be equal for all the other types (CurrentCulture, InvariantCulture, *IgnoreCase, etc.) as well.

I'm going to submit a PR soon which includes a fix for this anyway, but I just wanted to let people know/confirm that this wasn't intentional.

cc @jkotas @bbowyersmyth @GSPP

### PR DESCRIPTION
**Pull request by jamesqo**
Currently there are a lot of overloads of String.Compare that feature needless code duplication. This PR removes some of that, and in the process optimizes some overloads, as they're able to now short-circuit for things like reference equality where they didn't before.

### CODE REVIEW

**Comment by mikedn** [comment]
IMO there are way too many commits in this PR and they contain too many unnecessary and even unrelated changes. It would be nice to clean it up before review.

File: src/mscorlib/src/System/String.cs [hunk 1]
```diff
@@ -2087,59 +2105,42 @@ public static int Compare(String strA, String strB, StringComparison comparisonT
...
-            if (length < 0) {
-                throw new ArgumentOutOfRangeException("length",
-                                                      Environment.GetResourceString("ArgumentOutOfRange_NegativeLength"));
+                return strA == null ? -1 : 1;
             }
-
-            if (indexA < 0) {
-                throw new ArgumentOutOfRangeException("indexA",
-                                                     Environment.GetResourceString("ArgumentOutOfRange_Index"));
+
```
Comment (jasonwilliams200OK) [inline src/mscorlib/src/System/String.cs]: nit: extra spacing

File: src/mscorlib/src/System/String.cs [hunk 2]
```diff
@@ -2174,43 +2175,48 @@ public static int Compare(String strA, String strB, StringComparison comparisonT
...
-            if (!(value is String)) {
+            string other = value as string;
+
+            if (other == null)
+            {
                 throw new ArgumentException(Environment.GetResourceString("Arg_MustBeString"));
             }
-
-            return String.Compare(this,(String)value, StringComparison.CurrentCulture);
+
```
Comment (jasonwilliams200OK) [inline src/mscorlib/src/System/String.cs]: nit: extra spacing

**Comment by jasonwilliams200OK** [comment]
I agree with @mikedn; @jamesqo, despite the presence of [squash on merge](https://help.github.com/articles/about-pull-request-merge-squashing/) GitHub feature, since this change only affects one file, you can squash 26 commits to 1 and write the description on changes as bullet points in commit log. e.g. https://github.com/dotnet/coreclr/commit/6fe08a80542d496e657eaed4902bd33ebce4c6f8 (keeping column width <=50 chars for the first line and <= 72 chars for the rest of the lines, see https://github.com/libuv/libuv/blob/v1.x/CONTRIBUTING.md#commit and [The seven rules of a great git commit message](http://chris.beams.io/posts/git-commit/))

File: src/mscorlib/src/System/String.cs (hunk 2 above)
Comment (jasonwilliams200OK) [inline src/mscorlib/src/System/String.cs]: These extra whitespaces are only visible in the diff view: https://github.com/dotnet/coreclr/pull/6603/files

Comment (jamesqo) [inline src/mscorlib/src/System/String.cs]: @jasonwilliams200OK IMO I'm not sure that the whitespace changes matter all that much, as long as they're spaces and they're either aligned with the braces or the beginning of the line. GitHub has a trick where you can append ?w=1 to the end of the URL to eliminate the whitespace changes in a diff, e.g.:

https://github.com/dotnet/coreclr/pull/6603/files?w=1

Comment (jasonwilliams200OK) [inline src/mscorlib/src/System/String.cs]: @jamesqo, aware of that hack and agree that it doesn't matter, but i saw that in other places of this PR you have removed the extra whitespaces but these added extras at two other places.

**Comment by jamesqo** [comment]
@jkotas I have cleaned up the commit history, and I'm going to submit the non-trivial commits in another PR. This should be ready for review.

**Comment by jkotas** [comment]
LGTM - makes the implementation more compact. Thanks!

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by jamesqo** [comment]
Will port along with #6891

**Pull request merged**